### PR TITLE
JENKINS-53353 Do not cache results of View.getActions()

### PR DESCRIPTION
**Pull request by agentgonzo**
See [JENKINS-53353](https://issues.jenkins-ci.org/browse/JENKINS-53353).

Plugins that may affect the view, eg via TransientViewActionFactory
that get installed _after_ a `view` has been initally viewed
will not be able to add anything to the view if the results
of getActions() are cached.

### Proposed changelog entries

* Entry 1: Removed caching in Views - Plugins that affected the View, but are installed _after_ the initial rendering/viewing of the `View` will not be called again and do not contribute to the `View`.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
      * No existing tests present for the caching. Code is simplistic enough and removes functionality so unnecessary to add tests.
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

@reviewbybees @jtnord

### CODE REVIEW

**Comment by agentgonzo** [comment]
Trying to avoid premature optimisation :-). @jtnord indicated that it may not be saving much processing anyway, so I thought I'd put it up for review first. If people who understood the code far better than me said that going via an `ExtensionListListener` would be the best way to go, I'd then go via that route (which is what look like is happening!).

I'll start looking at implementing a listener now.

**Comment by jtnord** [comment]
@daniel-beck a cache is useless as by definition a `TransientActionViewFactory` should be able to decide to add an action for a specific view based on some other information that could be temproral based, or based on the view containing a job where the last build had failed.

The `TransientActionFactory` (Why we have a specialisation I do not know) does not perform any cachine so it is not clear at all why we would need to.

**Comment by agentgonzo** [comment]
@daniel-beck - do you think that the `TransientViewActionFactory` _should_ perform caching? Given the comments above, I don't understand the reasons why it should at the moment.

**Comment by jtnord** [comment]
> However the fact that the cache exists means someone thought it's a good idea, so I'd hope we don't perform a change like this blindly, potentially introducing (performance) regressions.

Given It has existed since it was [introduced](https://github.com/jenkinsci/jenkins/commit/97f5e8d8dfdb87c2c04fb912528021f7c60f4bb7) it is hard to qaulify the thinking (given it was not a specific change).  Only @huybrechts really knows and I doubt Tom can recall his thought process of 8 years +.

I would say go with this as it is a better fix (Factories can decided if they want to have an action based on the view state rather than have to always add the action and later hide it) and if it causes an issue this can be reverted and replaced by the narrower fix in #3610

**Comment by daniel-beck** [comment]
> I would say go with this as it is a better fix (Factories can decided if they want to have an action based on the view state rather than have to always add the action and later hide it) and if it causes an issue this can be reverted

Seems reasonable.

**Comment by agentgonzo** [comment]
Excellent - thanks both. Is it good to merge/release then? I'll leave #3610 hanging around for a fair amount of time in case this causes issues - you can then revert this and just merge #3610 if you need it.

**Comment by agentgonzo** [comment]
@daniel-beck - if you are OK with this fix, can you merge it please?

**Review by jtnord — APPROVED** [review]

File: core/src/main/java/hudson/model/View.java [hunk 1]
```diff
@@ -552,18 +549,11 @@ public String getSearchUrl() {
     	List<Action> result = new ArrayList<Action>();
     	result.addAll(getOwner().getViewActions());
     	synchronized (this) {
```
Comment (jtnord) [inline core/src/main/java/hudson/model/View.java]: @agentgonzo I don't think there is a need to keep this `synchronized` anymore is there?

Comment (agentgonzo) [reply core/src/main/java/hudson/model/View.java]: nope

File: core/src/main/java/hudson/model/View.java [hunk 2]
```diff
@@ -552,18 +549,11 @@ public String getSearchUrl() {
...
-    		if (transientActions == null) {
-                updateTransientActions();
-    		}
-    		result.addAll(transientActions);
+            result.addAll(TransientViewActionFactory.createAllFor(this));
     	}
     	return result;
     }
-    
-    public synchronized void updateTransientActions() {
```
Comment (jtnord) [inline core/src/main/java/hudson/model/View.java]: NB: this is a backwards incompatable API change (however a search of public sources in the jenkinsci org found no usage of it.

File: core/src/main/java/hudson/model/View.java [hunk 3]
```diff
@@ -172,8 +171,6 @@
      */
     protected boolean filterQueue;
     
-    protected transient List<Action> transientActions;
```
Comment (jtnord) [inline core/src/main/java/hudson/model/View.java]: NB: this is a backwards incompatible API change (however a search of public sources in the jenkinsci org found no usage of it.

**Comment by agentgonzo** [comment]
Build failures seem infra-related: `1) Error in custom provider, org.openqa.selenium.WebDriverException: Failed to connect to binary FirefoxBinary(/usr/bin/firefox) on port 7055;`

**Review by daniel-beck — APPROVED** [review]

**Comment by agentgonzo** [comment]
@oleg-nenashev - have you had time to look at this yet?

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
It breaks binary compatibility. Should retain a stub `updateTransientActions()` method which is deprecated & does nothing (and no need to synchronize it anymore)

**Comment by agentgonzo** [comment]
@oleg-nenashev - done

File: .mvn/jvm.config [hunk 4]
```diff
@@ -1 +0,0 @@
--Xmx800m
```
Comment (oleg-nenashev) [inline .mvn/jvm.config]: why?

Comment (agentgonzo) [reply .mvn/jvm.config]: committed in error (my local machine can't build with that in there for some reason)
I've replaced the commit

File: core/src/main/java/hudson/model/View.java (hunk 2 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/View.java]: fixed

**Review by oleg-nenashev — APPROVED** [review]

**Pull request merged**